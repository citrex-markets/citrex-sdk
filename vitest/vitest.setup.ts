import { vi } from 'vitest'
import createFetchMock from 'vitest-fetch-mock'

vi.spyOn(console, 'error').mockImplementation(() => {})
vi.spyOn(console, 'log').mockImplementation(() => {})

const fetchMocker = createFetchMock(vi)
fetchMocker.enableMocks()

Object.defineProperty(BigInt.prototype, 'toJSON', {
  get() {
    'use strict'
    return () => String(this)
  },
})

vi.stubEnv('CIAO_MAINNET_ADDRESS', '0x7461cFe1A4766146cAFce60F6907Ea657550670d')
vi.stubEnv('CIAO_TESTNET_ADDRESS', '0x0F571400ef7D2aEc68b29e58be3adCE1Bb27f33d')
vi.stubEnv('ORDER_DISPATCH_MAINNET_ADDRESS', '0x993543DC8BdFCba9fc7355d822108eF49dB6b9F9')
vi.stubEnv('ORDER_DISPATCH_TESTNET_ADDRESS', '0x24f4e9Db8225e6AE220FE89782E4A010aEB7bb14')
