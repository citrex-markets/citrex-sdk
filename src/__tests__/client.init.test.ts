import { beforeEach, describe, expect, it, vi } from 'vitest'

import CitrexSDK from 'src'
import { Environment } from 'src/enums'
import { privateKey } from 'vitest/utils'

const mocks = vi.hoisted(() => ({
  createPublicClient: vi.fn().mockReturnValue({}),
}))

vi.mock('viem', async () => {
  const actual = (await vi.importActual('viem')) as any

  return {
    ...actual,
    createPublicClient: mocks.createPublicClient,
  }
})

describe('The CitrexSDK', () => {
  beforeEach(() => {
    fetchMock.resetMocks()
  })

  it('should initialise correctly with only required params', () => {
    const Client = new CitrexSDK(privateKey)

    expect(Client).toMatchSnapshot()
  })

  it('should initialise correctly with all config parameters passed', () => {
    const Client = new CitrexSDK(privateKey, {
      debug: true,
      environment: Environment.TESTNET,
      rpc: 'https://test-rpc.quiknode.pro',
      subAccountId: 2,
    })

    expect(Client).toMatchSnapshot()
  })

  it('should POST the referral code if running on mainnet', async () => {
    fetchMock.mockResponse(JSON.stringify({ success: true }))

    new CitrexSDK(privateKey, { environment: Environment.MAINNET })

    await vi.waitFor(() => {
      if (!fetchMock.mock.calls.length) throw new Error('Awaiting call...')
    })

    expect(fetchMock.mock.lastCall).toMatchInlineSnapshot(`
      [
        "https://api.citrex.markets/v1/vault/referral",
        {
          "body": "{"account":"0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8","code":"kickflip"}",
          "method": "POST",
        },
      ]
    `)
  })

  it('should handle referral failures gracefully', async () => {
    fetchMock.mockReject(new Error('An unknown error occurred'))

    const Client = new CitrexSDK(privateKey, { debug: true, environment: Environment.MAINNET })

    await vi.waitFor(() => {
      if (Client.logs.length !== 3) throw new Error('Awaiting call...')
    })

    expect(Client.logs[2]).toEqual({ msg: 'Call failed, ignoring.' })
  })
})
