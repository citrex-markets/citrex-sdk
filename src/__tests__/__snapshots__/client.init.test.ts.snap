// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The CitrexSDK > should initialise correctly with all config parameters passed 1`] = `
CitrexSDK {
  "account": {
    "address": "0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8",
    "experimental_signAuthorization": [Function],
    "nonceManager": undefined,
    "publicKey": "0x0422eea015a8664523fce2eed5de514ad98a29724755c488f3d11d913d5671c881bc2d7b1e0d17389604ef141a96830b14917b034b7621579d1159d694dd231798",
    "sign": [Function],
    "signMessage": [Function],
    "signTransaction": [Function],
    "signTypedData": [Function],
    "source": "privateKey",
    "type": "local",
  },
  "apiUrl": "https://api.staging.citrex.markets/v1",
  "calculateMarginRequirement": [Function],
  "cancelAndReplaceOrder": [Function],
  "cancelOpenOrdersForProduct": [Function],
  "cancelOrder": [Function],
  "cancelOrders": [Function],
  "chain": {
    "blockExplorers": {
      "default": {
        "name": "Seitrace",
        "url": "https://seitrace.com",
      },
    },
    "fees": undefined,
    "formatters": undefined,
    "id": 1328,
    "name": "Sei Testnet",
    "nativeCurrency": {
      "decimals": 18,
      "name": "Sei",
      "symbol": "SEI",
    },
    "rpcUrls": {
      "default": {
        "http": [
          "https://evm-rpc-testnet.sei-apis.com",
        ],
        "webSocket": [
          "wss://evm-ws-testnet.sei-apis.com",
        ],
      },
    },
    "serializers": undefined,
    "testnet": true,
  },
  "ciaoAddress": "0x0F571400ef7D2aEc68b29e58be3adCE1Bb27f33d",
  "deposit": [Function],
  "domain": {
    "chainId": 1328n,
    "name": "ciao",
    "verifyingContract": "0x24f4e9Db8225e6AE220FE89782E4A010aEB7bb14",
    "version": "0.0.0",
  },
  "environment": "testnet",
  "getAccountHealth": [Function],
  "getKlines": [Function],
  "getOrderBook": [Function],
  "getProduct": [Function],
  "getProducts": [Function],
  "getServerTime": [Function],
  "getTickers": [Function],
  "getTradeHistory": [Function],
  "listBalances": [Function],
  "listOpenOrders": [Function],
  "listPositions": [Function],
  "logs": [
    {
      "msg": "Debug mode enabled",
    },
  ],
  "marginAssets": {
    "USDC": "0x79A59c326C715AC2d31C169C85d1232319E341ce",
  },
  "placeOrder": [Function],
  "placeOrders": [Function],
  "publicClient": {},
  "rpc": "https://test-rpc.quiknode.pro",
  "subAccountId": 2,
  "verifierAddress": "0x24f4e9Db8225e6AE220FE89782E4A010aEB7bb14",
  "withdraw": [Function],
}
`;

exports[`The CitrexSDK > should initialise correctly with only required params 1`] = `
CitrexSDK {
  "account": {
    "address": "0xb47B0b1e44B932Ae9Bb01817E7010A553A965Ea8",
    "experimental_signAuthorization": [Function],
    "nonceManager": undefined,
    "publicKey": "0x0422eea015a8664523fce2eed5de514ad98a29724755c488f3d11d913d5671c881bc2d7b1e0d17389604ef141a96830b14917b034b7621579d1159d694dd231798",
    "sign": [Function],
    "signMessage": [Function],
    "signTransaction": [Function],
    "signTypedData": [Function],
    "source": "privateKey",
    "type": "local",
  },
  "apiUrl": "https://api.staging.citrex.markets/v1",
  "calculateMarginRequirement": [Function],
  "cancelAndReplaceOrder": [Function],
  "cancelOpenOrdersForProduct": [Function],
  "cancelOrder": [Function],
  "cancelOrders": [Function],
  "chain": {
    "blockExplorers": {
      "default": {
        "name": "Seitrace",
        "url": "https://seitrace.com",
      },
    },
    "fees": undefined,
    "formatters": undefined,
    "id": 1328,
    "name": "Sei Testnet",
    "nativeCurrency": {
      "decimals": 18,
      "name": "Sei",
      "symbol": "SEI",
    },
    "rpcUrls": {
      "default": {
        "http": [
          "https://evm-rpc-testnet.sei-apis.com",
        ],
        "webSocket": [
          "wss://evm-ws-testnet.sei-apis.com",
        ],
      },
    },
    "serializers": undefined,
    "testnet": true,
  },
  "ciaoAddress": "0x0F571400ef7D2aEc68b29e58be3adCE1Bb27f33d",
  "deposit": [Function],
  "domain": {
    "chainId": 1328n,
    "name": "ciao",
    "verifyingContract": "0x24f4e9Db8225e6AE220FE89782E4A010aEB7bb14",
    "version": "0.0.0",
  },
  "environment": "testnet",
  "getAccountHealth": [Function],
  "getKlines": [Function],
  "getOrderBook": [Function],
  "getProduct": [Function],
  "getProducts": [Function],
  "getServerTime": [Function],
  "getTickers": [Function],
  "getTradeHistory": [Function],
  "listBalances": [Function],
  "listOpenOrders": [Function],
  "listPositions": [Function],
  "logs": [],
  "marginAssets": {
    "USDC": "0x79A59c326C715AC2d31C169C85d1232319E341ce",
  },
  "placeOrder": [Function],
  "placeOrders": [Function],
  "publicClient": {},
  "rpc": "https://evm-rpc-testnet.sei-apis.com",
  "subAccountId": 1,
  "verifierAddress": "0x24f4e9Db8225e6AE220FE89782E4A010aEB7bb14",
  "withdraw": [Function],
}
`;
